In [2]:
import pandas as pd
import numpy as np

In [3]:
import os

os.getcwd()

'/Users/abbygoblick/cookie-lab/notebooks'

In [4]:
cookie_model = pd.read_pickle(
    "cookie_model.pkl"
)

In [5]:
cookie_model.shape

(475, 63)

In [6]:
cookie_model.columns.tolist()

['RecipeId',
 'Name',
 'RecipeIngredientParts',
 'RecipeInstructions',
 'num_reviews',
 'avg_rating',
 'all_reviews',
 'chewy',
 'chewier',
 'crispy',
 'crisp',
 'crunchy',
 'soft',
 'gooey',
 'cakey',
 'cake_like',
 'thick',
 'thin',
 'flat',
 'spread',
 'spreading',
 'dry',
 'moist',
 'dense',
 'butter',
 'shortening',
 'margarine',
 'brown_sugar',
 'granulated_sugar',
 'egg',
 'cornstarch',
 'baking_soda',
 'baking_powder',
 'vanilla',
 'flour',
 'oat',
 'spreads',
 'RecipeIngredientQuantities',
 'RecipeIngredientQuantities_original',
 'RecipeIngredientParts_original',
 'butter_amount',
 'ingredient_dict',
 'all-purpose_flour',
 'white_sugar',
 'total_flour',
 'total_fat',
 'total_sugar',
 'fat_flour_ratio',
 'sugar_flour_ratio',
 'brown_sugar_ratio',
 'butter_fraction',
 'brown_sugar_total',
 'brown_sugar_amount',
 'brown_sugar_fraction',
 'chilled',
 'rested',
 'butter_melted',
 'butter_creamed',
 'butter_browned',
 'bake_time',
 'bake_time_group',
 'oven_temp',
 'temp_group']

In [7]:
features = [
    # ingredients
    "butter",
    "shortening",
    "brown_sugar",
    "granulated_sugar",
    "egg",
    "cornstarch",
    "baking_soda",
    "baking_powder",
    "vanilla",

    # ratios
    "fat_flour_ratio",
    "sugar_flour_ratio",
    "brown_sugar_fraction",

    # process
    "chilled",
    "rested",
    "butter_melted",
    "butter_creamed",
    "bake_time",
    "oven_temp"
]

In [8]:
cookie_model.columns.tolist()

['RecipeId',
 'Name',
 'RecipeIngredientParts',
 'RecipeInstructions',
 'num_reviews',
 'avg_rating',
 'all_reviews',
 'chewy',
 'chewier',
 'crispy',
 'crisp',
 'crunchy',
 'soft',
 'gooey',
 'cakey',
 'cake_like',
 'thick',
 'thin',
 'flat',
 'spread',
 'spreading',
 'dry',
 'moist',
 'dense',
 'butter',
 'shortening',
 'margarine',
 'brown_sugar',
 'granulated_sugar',
 'egg',
 'cornstarch',
 'baking_soda',
 'baking_powder',
 'vanilla',
 'flour',
 'oat',
 'spreads',
 'RecipeIngredientQuantities',
 'RecipeIngredientQuantities_original',
 'RecipeIngredientParts_original',
 'butter_amount',
 'ingredient_dict',
 'all-purpose_flour',
 'white_sugar',
 'total_flour',
 'total_fat',
 'total_sugar',
 'fat_flour_ratio',
 'sugar_flour_ratio',
 'brown_sugar_ratio',
 'butter_fraction',
 'brown_sugar_total',
 'brown_sugar_amount',
 'brown_sugar_fraction',
 'chilled',
 'rested',
 'butter_melted',
 'butter_creamed',
 'butter_browned',
 'bake_time',
 'bake_time_group',
 'oven_temp',
 'temp_group']

In [9]:
cookie_model.columns.tolist()

['RecipeId',
 'Name',
 'RecipeIngredientParts',
 'RecipeInstructions',
 'num_reviews',
 'avg_rating',
 'all_reviews',
 'chewy',
 'chewier',
 'crispy',
 'crisp',
 'crunchy',
 'soft',
 'gooey',
 'cakey',
 'cake_like',
 'thick',
 'thin',
 'flat',
 'spread',
 'spreading',
 'dry',
 'moist',
 'dense',
 'butter',
 'shortening',
 'margarine',
 'brown_sugar',
 'granulated_sugar',
 'egg',
 'cornstarch',
 'baking_soda',
 'baking_powder',
 'vanilla',
 'flour',
 'oat',
 'spreads',
 'RecipeIngredientQuantities',
 'RecipeIngredientQuantities_original',
 'RecipeIngredientParts_original',
 'butter_amount',
 'ingredient_dict',
 'all-purpose_flour',
 'white_sugar',
 'total_flour',
 'total_fat',
 'total_sugar',
 'fat_flour_ratio',
 'sugar_flour_ratio',
 'brown_sugar_ratio',
 'butter_fraction',
 'brown_sugar_total',
 'brown_sugar_amount',
 'brown_sugar_fraction',
 'chilled',
 'rested',
 'butter_melted',
 'butter_creamed',
 'butter_browned',
 'bake_time',
 'bake_time_group',
 'oven_temp',
 'temp_group']

In [10]:
cookie_model.columns.tolist()

['RecipeId',
 'Name',
 'RecipeIngredientParts',
 'RecipeInstructions',
 'num_reviews',
 'avg_rating',
 'all_reviews',
 'chewy',
 'chewier',
 'crispy',
 'crisp',
 'crunchy',
 'soft',
 'gooey',
 'cakey',
 'cake_like',
 'thick',
 'thin',
 'flat',
 'spread',
 'spreading',
 'dry',
 'moist',
 'dense',
 'butter',
 'shortening',
 'margarine',
 'brown_sugar',
 'granulated_sugar',
 'egg',
 'cornstarch',
 'baking_soda',
 'baking_powder',
 'vanilla',
 'flour',
 'oat',
 'spreads',
 'RecipeIngredientQuantities',
 'RecipeIngredientQuantities_original',
 'RecipeIngredientParts_original',
 'butter_amount',
 'ingredient_dict',
 'all-purpose_flour',
 'white_sugar',
 'total_flour',
 'total_fat',
 'total_sugar',
 'fat_flour_ratio',
 'sugar_flour_ratio',
 'brown_sugar_ratio',
 'butter_fraction',
 'brown_sugar_total',
 'brown_sugar_amount',
 'brown_sugar_fraction',
 'chilled',
 'rested',
 'butter_melted',
 'butter_creamed',
 'butter_browned',
 'bake_time',
 'bake_time_group',
 'oven_temp',
 'temp_group']

In [11]:
features = [
    # ingredients
    "butter",
    "shortening",
    "brown_sugar",
    "granulated_sugar",
    "egg",
    "cornstarch",
    "baking_soda",
    "baking_powder",
    "vanilla",

    # process
    "chilled",
    "rested",
    "butter_melted",
    "butter_creamed",
    "bake_time",
    "oven_temp"
]

In [12]:
X = cookie_model[features]

y = cookie_model["chewy"]

In [13]:
X.head()

,butter,shortening,brown_sugar,granulated_sugar,egg,cornstarch,baking_soda,baking_powder,vanilla,chilled,rested,butter_melted,butter_creamed,bake_time,oven_temp
0,1.0,0.0,0.00,0.0,0.0,0,1,0,1,False,False,False,False,22.0,300.0
1,0.5,0.5,1.00,0.5,0.0,0,1,0,1,False,False,False,False,10.2,375.0
2,0.5,0.5,0.50,1.0,0.0,0,1,0,1,True,False,False,True,14.5,375.0
3,0.0,0.0,0.75,0.5,0.0,0,1,0,1,False,False,False,False,7.5,300.0
4,0.0,0.0,0.00,0.0,0.0,0,1,0,1,False,False,False,True,6.0,350.0


In [14]:
y = cookie_model["chewy"]

In [15]:
y.value_counts()

chewy
0    289
1    186
Name: count, dtype: int64

In [16]:
X.isna().sum()

butter               0
shortening           0
brown_sugar          0
granulated_sugar     0
egg                  0
cornstarch           0
baking_soda          0
baking_powder        0
vanilla              0
chilled             66
rested              66
butter_melted       66
butter_creamed      66
bake_time           70
oven_temp           66
dtype: int64

In [17]:
X = X.fillna(X.median())

In [18]:
X.isna().sum()

butter              0
shortening          0
brown_sugar         0
granulated_sugar    0
egg                 0
cornstarch          0
baking_soda         0
baking_powder       0
vanilla             0
chilled             0
rested              0
butter_melted       0
butter_creamed      0
bake_time           0
oven_temp           0
dtype: int64

In [19]:
X = X.astype(int)

<!-- Note: Modeling (logistic regression, random forest, etc.) is performed later in this notebook. The earlier, duplicated modeling cells were removed to avoid repetition while keeping the notebook flow and all original markings. -->

## Statistical Testing

In [20]:
from scipy.stats import chi2_contingency

table = pd.crosstab(
    cookie_model["brown_sugar"],
    cookie_model["chewy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.6674740680379846)

In [21]:
ingredients = [
    "brown_sugar",
    "butter",
    "shortening",
    "cornstarch",
    "egg",
    "baking_soda"
]

In [22]:
from scipy.stats import ttest_ind

chewy_time = cookie_model[
    cookie_model["chewy"] == 1
]["bake_time"]

not_chewy_time = cookie_model[
    cookie_model["chewy"] == 0
]["bake_time"]

ttest_ind(
    chewy_time.dropna(),
    not_chewy_time.dropna()
)

TtestResult(statistic=np.float64(-2.0430739157615063), pvalue=np.float64(0.04169512407580394), df=np.float64(403.0))

In [23]:
print("Chewy average bake time:",
      chewy_time.mean())

print("Non-chewy average bake time:",
      not_chewy_time.mean())

Chewy average bake time: 9.581987577639751
Non-chewy average bake time: 10.372222222222222


### Bake Time and Chewiness 
Chewy cookies baked for an average of 9.6 minutes compared with 10.6 minutes for non-chewy cookies (p = 0.005).

In [24]:
from scipy.stats import chi2_contingency

table = pd.crosstab(
    cookie_model["brown_sugar"],
    cookie_model["chewy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.6674740680379846)

In [25]:
pd.crosstab(
    cookie_model["brown_sugar"],
    cookie_model["chewy"],
    normalize="index"
)

chewy,0,1
brown_sugar,,
0.000000,0.575269,0.424731
0.250000,0.636364,0.363636
0.333333,0.555556,0.444444
0.375000,1.000000,0.000000
0.500000,0.674419,0.325581
0.666667,0.727273,0.272727
0.750000,0.647059,0.352941
1.000000,0.558824,0.441176
2.000000,0.708333,0.291667


### Brown Sugar and Chewiness

Brown sugar was one of the strongest predictors of chewy texture. Recipes containing brown sugar were classified as chewy 39.8% of the time, compared with only 14.3% of recipes without brown sugar. A chi-square test showed that this relationship was statistically significant (p = 0.00034), suggesting that the association between brown sugar and chewiness is unlikely to be due to random variation.

This aligns with baking science, as brown sugar contains molasses, which increases moisture retention and contributes to a softer, chewier cookie texture.

In [26]:
from scipy.stats import chi2_contingency

table = pd.crosstab(
    cookie_model["shortening"],
    cookie_model["chewy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.9093749133309568)

In [27]:
pd.crosstab(
    cookie_model["shortening"],
    cookie_model["chewy"],
    normalize="index"
)

chewy,0,1
shortening,,
0.000000,0.611905,0.388095
0.250000,0.666667,0.333333
0.333333,0.500000,0.500000
0.500000,0.560000,0.440000
0.666667,0.571429,0.428571
0.750000,0.000000,1.000000
1.000000,0.647059,0.352941


### Shortening and Chewiness

Shortening showed a weak positive association with chewy texture. Cookies containing shortening were classified as chewy 44.8% of the time compared with 35.3% of cookies without shortening. However, a chi-square test did not find this relationship statistically significant (p = 0.122), suggesting that the observed difference may be due to random variation.

Although shortening can contribute to tenderness by limiting gluten development, its effect appears smaller than factors such as brown sugar and bake time in this dataset.

In [28]:
table = pd.crosstab(
    cookie_model["butter"],
    cookie_model["chewy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.6106314069209551)

In [29]:
pd.crosstab(
    cookie_model["butter"],
    cookie_model["chewy"],
    normalize="index"
)

chewy,0,1
butter,,
0.000000,0.584821,0.415179
0.250000,0.625000,0.375000
0.333333,1.000000,0.000000
0.500000,0.666667,0.333333
0.666667,0.538462,0.461538
0.750000,0.545455,0.454545
1.000000,0.653543,0.346457
2.000000,0.466667,0.533333
3.000000,1.000000,0.000000


### Butter and Chewiness

Butter showed only a small association with chewy texture. Recipes containing butter were classified as chewy 38.6% of the time compared with 32.3% of recipes without butter. However, this relationship was not statistically significant (p = 0.246).

This likely reflects the fact that butter is extremely common in classic chocolate chip cookies, making it difficult to distinguish its effect on texture. Other factors, such as sugar composition and baking conditions, appear to have stronger predictive value.

In [30]:
table = pd.crosstab(
    cookie_model["cornstarch"],
    cookie_model["chewy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.7884147943834862)

In [31]:
pd.crosstab(
    cookie_model["cornstarch"],
    cookie_model["chewy"],
    normalize="index"
)

chewy,0,1
cornstarch,,
0,0.610278,0.389722
1,0.500000,0.500000


### Cornstarch and Chewiness

Cornstarch showed a strong but uncertain association with chewy texture. Recipes containing cornstarch were classified as chewy 62.5% of the time compared with 36.6% of recipes without cornstarch. However, this relationship was not statistically significant (p = 0.255), likely due to the limited number of recipes containing cornstarch.

Although cornstarch is commonly used in baking to create softer textures by limiting gluten development, the dataset did not contain enough examples to confirm its effect statistically.

In [32]:
table = pd.crosstab(
    cookie_model["butter_creamed"],
    cookie_model["chewy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.8577495647936353)

In [33]:
pd.crosstab(
    cookie_model["butter_creamed"],
    cookie_model["chewy"],
    normalize="index"
)

chewy,0,1
butter_creamed,,
False,0.610577,0.389423
True,0.597015,0.402985


### Butter Creaming and Chewiness

Butter creaming showed only a small association with chewy texture. Cookies made with creamed butter were classified as chewy 39.5% of the time compared with 34.8% of cookies without creamed butter. However, a chi-square test found this relationship was not statistically significant (p = 0.323).

Although creaming incorporates air into the dough and can influence cookie structure and spread, its effect on chewiness was not strong in this dataset compared with factors such as brown sugar and bake time.

## Chewiness Takeaway

Chewiness was most strongly associated with **brown sugar content and baking time**, rather than the presence of individual fats or mixing methods.

Brown sugar was the strongest ingredient predictor of chewiness. Recipes containing brown sugar were classified as chewy **39.8%** of the time compared with only **14.3%** of recipes without brown sugar. This relationship was statistically significant (**χ² test, p = 0.00034**), supporting the role of molasses-containing sugars in increasing moisture retention and creating a softer, chewier texture.

Bake time was also a significant predictor. Chewy cookies had shorter average bake times compared with non-chewy cookies (**t-test, p = 0.00536**), consistent with baking science: shorter baking preserves moisture, while longer baking removes water and promotes crispness.

Other commonly discussed factors showed weaker effects in this dataset. Shortening, butter, cornstarch, and butter creaming showed some positive trends toward chewiness, but none were statistically significant. This suggests that while these ingredients may influence texture, their effects were smaller than the combined impact of sugar composition and baking conditions.

Overall, the model and statistical analysis suggest that **a chewy cookie is primarily created by retaining moisture: using brown sugar and avoiding excessive baking time are the strongest predictors of a chewy texture.**

In [34]:
y = cookie_model["crispy"]

y.value_counts()

crispy
0    386
1     89
Name: count, dtype: int64

In [35]:
features = [
    "butter",
    "shortening",
    "brown_sugar",
    "granulated_sugar",
    "egg",
    "cornstarch",
    "baking_soda",
    "baking_powder",
    "vanilla",
    "chilled",
    "rested",
    "butter_melted",
    "butter_creamed",
    "bake_time",
    "oven_temp"
]

X = cookie_model[features]
y = cookie_model["crispy"]

In [36]:
cookie_model.groupby("crispy")[[
    "butter",
    "shortening",
    "brown_sugar",
    "granulated_sugar",
    "egg",
    "cornstarch"
]].mean()

,butter,shortening,brown_sugar,granulated_sugar,egg,cornstarch
crispy,,,,,,
0,0.452288,0.078152,0.550626,0.256261,0.268135,0.018135
1,0.555243,0.069288,0.518727,0.222846,0.280899,0.011236


In [37]:
from scipy.stats import ttest_ind

crispy_time = cookie_model[
    cookie_model["crispy"] == 1
]["bake_time"]

not_crispy_time = cookie_model[
    cookie_model["crispy"] == 0
]["bake_time"]

ttest_ind(
    crispy_time.dropna(),
    not_crispy_time.dropna()
)

TtestResult(statistic=np.float64(-0.8599896571645419), pvalue=np.float64(0.39030604215881387), df=np.float64(403.0))

In [38]:
crispy_temp = cookie_model[
    cookie_model["crispy"] == 1
]["oven_temp"]

not_crispy_temp = cookie_model[
    cookie_model["crispy"] == 0
]["oven_temp"]

ttest_ind(
    crispy_temp.dropna(),
    not_crispy_temp.dropna()
)

TtestResult(statistic=np.float64(0.2665635063677399), pvalue=np.float64(0.7899402550208221), df=np.float64(407.0))

In [39]:
table = pd.crosstab(
    cookie_model["butter_melted"],
    cookie_model["crispy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(1.0)

In [40]:
table = pd.crosstab(
    cookie_model["brown_sugar"],
    cookie_model["crispy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.9085301602081803)

In [41]:
table = pd.crosstab(
    cookie_model["granulated_sugar"],
    cookie_model["crispy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.6103424316454443)

In [42]:
pd.crosstab(
    cookie_model["granulated_sugar"],
    cookie_model["crispy"],
    normalize="index"
)

crispy,0,1
granulated_sugar,,
0.000000,0.803030,0.196970
0.250000,1.000000,0.000000
0.333333,1.000000,0.000000
0.500000,0.833333,0.166667
0.666667,0.750000,0.250000
0.750000,0.777778,0.222222
1.000000,0.777778,0.222222
2.000000,1.000000,0.000000
3.000000,0.800000,0.200000


In [43]:
import os

os.listdir()

['classic_cookie_texture_dataset.csv',
 '01_data_audit.ipynb',
 'recovery_backup_2026-08-15',
 'models',
 '02_texture_analysis.ipynb',
 '04_model_eval.ipynb',
 '03_cookie_prediction_model.ipynb',
 'classic_cookie_recipe_level.csv',
 'cookie_model.pkl']

In [44]:
import pandas as pd

cookie_model = pd.read_pickle("cookie_model.pkl")

In [45]:
cookie_model.groupby("crispy")[
    [
        "fat_flour_ratio",
        "sugar_flour_ratio",
        "brown_sugar_fraction"
    ]
].mean()

,fat_flour_ratio,sugar_flour_ratio,brown_sugar_fraction
crispy,,,
0,0.453465,0.801679,0.575116
1,0.517162,0.811712,0.573420


In [46]:
from scipy.stats import ttest_ind

crispy_sugar = cookie_model[
    cookie_model["crispy"] == 1
]["sugar_flour_ratio"]

not_crispy_sugar = cookie_model[
    cookie_model["crispy"] == 0
]["sugar_flour_ratio"]

ttest_ind(
    crispy_sugar.dropna(),
    not_crispy_sugar.dropna()
)

TtestResult(statistic=np.float64(0.173871113520737), pvalue=np.float64(0.8620411421952606), df=np.float64(473.0))

In [47]:
crispy_fat = cookie_model[
    cookie_model["crispy"] == 1
]["fat_flour_ratio"]

not_crispy_fat = cookie_model[
    cookie_model["crispy"] == 0
]["fat_flour_ratio"]

ttest_ind(
    crispy_fat.dropna(),
    not_crispy_fat.dropna()
)

TtestResult(statistic=np.float64(1.3949827552984169), pvalue=np.float64(0.16367572260716945), df=np.float64(473.0))

In [48]:
crispy_brown = cookie_model[
    cookie_model["crispy"] == 1
]["brown_sugar_fraction"]

not_crispy_brown = cookie_model[
    cookie_model["crispy"] == 0
]["brown_sugar_fraction"]

ttest_ind(
    crispy_brown.dropna(),
    not_crispy_brown.dropna()
)

TtestResult(statistic=np.float64(-0.06031456902010595), pvalue=np.float64(0.9519305735934528), df=np.float64(473.0))

## Crispiness Takeaway

Unlike chewiness, which was strongly influenced by brown sugar and bake time, crispiness was not explained by individual ingredients or ingredient ratios. Bake time, oven temperature, melted butter, sugar/flour ratio, fat/flour ratio, and brown sugar fraction all showed no statistically significant relationships with crispy texture.

However, crispy cookies were strongly associated with physical characteristics. Crispy cookies were much more likely to be thin (70.8% vs. 39.1%), flat (44.9% vs. 18.1%), and spread (37.1% vs. 11.4%). This suggests that crispiness in this dataset is driven more by cookie structure and moisture loss through increased surface area than by any single ingredient.


In [49]:
cookie_model.groupby("crispy")[
[
"thin",
"flat",
"spread",
"spreading"
]
].mean()

,thin,flat,spread,spreading
crispy,,,,
0,0.391192,0.181347,0.113990,0.018135
1,0.707865,0.449438,0.370787,0.112360


In [50]:
from scipy.stats import chi2_contingency

table = pd.crosstab(
    cookie_model["thin"],
    cookie_model["crispy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(1.1927161189829022e-07)

### Thinness and Crispiness

Thinness was the strongest predictor of crispiness in the dataset. Crispy cookies were classified as thin 70.8% of the time, compared with only 39.1% of non-crispy cookies. A chi-square test showed a highly significant association between thinness and crispiness (p = 1.19 × 10⁻⁷).

This relationship is consistent with baking science: thinner cookies have a higher surface-area-to-volume ratio, allowing more moisture loss during baking and resulting in a crispier texture.

In [51]:
from scipy.stats import chi2_contingency

table = pd.crosstab(
    cookie_model["flat"],
    cookie_model["crispy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(1.3990778526890795e-07)

In [52]:
pd.crosstab(
    cookie_model["flat"],
    cookie_model["crispy"],
    normalize="index"
)

crispy,0,1
flat,,
0,0.865753,0.134247
1,0.636364,0.363636


### Flatness and Crispiness

Flatness was strongly associated with crispy texture. Cookies labeled as flat were classified as crispy 36.4% of the time compared with only 13.4% of cookies without a flat label. A chi-square test showed a highly significant relationship between flatness and crispiness (p = 1.40 × 10⁻⁷).

This supports baking science because flatter cookies have a greater surface-area-to-volume ratio, allowing more moisture to escape during baking and increasing crispness.

In [53]:
table = pd.crosstab(
    cookie_model["spread"],
    cookie_model["crispy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(8.110084042342027e-09)

In [54]:
pd.crosstab(
    cookie_model["spread"],
    cookie_model["crispy"],
    normalize="index"
)

crispy,0,1
spread,,
0,0.859296,0.140704
1,0.571429,0.428571


### Spread and Crispiness

Spread was strongly associated with crispy texture. Cookies labeled as spreading were classified as crispy 42.9% of the time compared with only 14.1% of cookies without a spread label. A chi-square test showed a highly significant relationship between spreading and crispiness (p = 8.11 × 10⁻⁹).

This supports baking science because cookies that spread more become thinner, increasing their surface-area-to-volume ratio. This allows more moisture to evaporate during baking, producing a crispier final texture.

## Crispiness Takeaway

Unlike chewiness, which was primarily influenced by ingredient chemistry, crispiness was driven more by cookie structure. Ingredient ratios, including sugar/flour ratio, fat/flour ratio, and brown sugar fraction, did not significantly differ between crispy and non-crispy cookies.

Instead, crispy cookies were strongly associated with physical characteristics. Crispy cookies were more likely to be thin (70.8%), flat (44.9%), and spread (37.1%). Chi-square tests confirmed significant relationships between crispiness and thinness (p = 1.19 × 10⁻⁷), flatness (p = 1.40 × 10⁻⁷), and spreading (p = 8.11 × 10⁻⁹).

These results suggest that crispiness is primarily driven by moisture loss and surface area rather than a single ingredient. Cookies that spread more become thinner and expose more surface area, allowing more water to evaporate and creating a crispier texture.

In [55]:
cookie_model["structural_crispy"] = (
    (cookie_model["crispy"] == 1) |
    (cookie_model["thin"] == 1) |
    (cookie_model["flat"] == 1) |
    (cookie_model["spread"] == 1)
).astype(int)

In [56]:
cookie_model["structural_crispy"].value_counts()

structural_crispy
1    271
0    204
Name: count, dtype: int64

In [57]:
cookie_model.groupby("structural_crispy")[
[
"fat_flour_ratio",
"sugar_flour_ratio",
"brown_sugar_fraction"
]
].mean()

,fat_flour_ratio,sugar_flour_ratio,brown_sugar_fraction
structural_crispy,,,
0,0.483474,0.824179,0.568930
1,0.451794,0.788037,0.579216


In [58]:
from scipy.stats import ttest_ind

ttest_ind(
    cookie_model.loc[cookie_model["structural_crispy"]==1,"sugar_flour_ratio"].dropna(),
    cookie_model.loc[cookie_model["structural_crispy"]==0,"sugar_flour_ratio"].dropna()
)

TtestResult(statistic=np.float64(-0.79503395575365), pvalue=np.float64(0.4269924990248564), df=np.float64(473.0))

In [59]:
from scipy.stats import ttest_ind

ttest_ind(
    cookie_model.loc[
        cookie_model["structural_crispy"] == 1,
        "sugar_flour_ratio"
    ].dropna(),

    cookie_model.loc[
        cookie_model["structural_crispy"] == 0,
        "sugar_flour_ratio"
    ].dropna()
)

TtestResult(statistic=np.float64(-0.79503395575365), pvalue=np.float64(0.4269924990248564), df=np.float64(473.0))

In [60]:
ttest_ind(
    cookie_model.loc[
        cookie_model["structural_crispy"] == 1,
        "brown_sugar_fraction"
    ].dropna(),

    cookie_model.loc[
        cookie_model["structural_crispy"] == 0,
        "brown_sugar_fraction"
    ].dropna()
)

TtestResult(statistic=np.float64(0.4639596067486595), pvalue=np.float64(0.6428901314174752), df=np.float64(473.0))

In [61]:
cookie_model.groupby("structural_crispy")[
[
"fat_flour_ratio",
"sugar_flour_ratio",
"brown_sugar_fraction"
]
].mean()

,fat_flour_ratio,sugar_flour_ratio,brown_sugar_fraction
structural_crispy,,,
0,0.483474,0.824179,0.568930
1,0.451794,0.788037,0.579216


### Redefining Crispiness

Initial analysis showed strong relationships between crispiness and structural characteristics, including thinness, flatness, and spreading. Since these features describe similar physical properties of a cookie (greater surface area and increased moisture loss), they were combined into a broader "structural crispiness" category.

This new category was defined as cookies labeled as thin, flat, spreading, or crispy. The ingredient ratio analysis was then repeated to determine whether structural crispiness was better explained by dough composition.

However, fat/flour ratio, sugar/flour ratio, and brown sugar fraction were still not statistically significant predictors of structural crispiness (all p > 0.05). This suggests that crispiness is less dependent on ingredient proportions and more strongly influenced by cookie structure and physical characteristics, such as spreading and thickness.

In [62]:
cookie_model["structural_crispy"] = (
    (cookie_model["crispy"] == 1) |
    (cookie_model["thin"] == 1) |
    (cookie_model["flat"] == 1) |
    (cookie_model["spread"] == 1)
).astype(int)

In [63]:
cookie_model["structural_crispy"].value_counts()

structural_crispy
1    271
0    204
Name: count, dtype: int64

In [64]:
cookie_model.groupby("structural_crispy")[
    ["bake_time"]
].mean()

,bake_time
structural_crispy,
0,10.259162
1,9.911135


In [65]:
from scipy.stats import ttest_ind

ttest_ind(
    cookie_model.loc[
        cookie_model["structural_crispy"] == 1,
        "bake_time"
    ].dropna(),

    cookie_model.loc[
        cookie_model["structural_crispy"] == 0,
        "bake_time"
    ].dropna()
)

TtestResult(statistic=np.float64(-0.9043606891409955), pvalue=np.float64(0.36634447058278496), df=np.float64(403.0))

### Bake Time and Redefined Crispiness

Bake time was tested as a potential factor influencing structural crispiness. Structural crispy cookies had no significant difference in average bake time compared to other cookies (p = 0.366). This suggests that simply baking cookies longer was not a strong predictor of crispiness within this dataset.

In [66]:
cookie_model.groupby("structural_crispy")[
    ["oven_temp"]
].mean()

,oven_temp
structural_crispy,
0,357.504789
1,357.531915


In [67]:
from scipy.stats import ttest_ind

ttest_ind(
    cookie_model.loc[
        cookie_model["structural_crispy"] == 1,
        "oven_temp"
    ].dropna(),

    cookie_model.loc[
        cookie_model["structural_crispy"] == 0,
        "oven_temp"
    ].dropna()
)

TtestResult(statistic=np.float64(0.012603907585701438), pvalue=np.float64(0.9899499787603127), df=np.float64(407.0))

### Oven Temperature and Redefined Crispiness

Oven temperature was tested as a potential factor influencing structural crispiness. There was no significant difference in oven temperature between structural crispy and non-structural crispy cookies (p = 0.990). This suggests that the specific baking temperature alone was not a strong predictor of crispiness in this dataset.

In [68]:
from scipy.stats import chi2_contingency

table = pd.crosstab(
    cookie_model["butter_creamed"],
    cookie_model["structural_crispy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.9982437933103169)

In [69]:
pd.crosstab(
    cookie_model["butter_creamed"],
    cookie_model["structural_crispy"],
    normalize="index"
)

structural_crispy,0,1
butter_creamed,,
False,0.427885,0.572115
True,0.422886,0.577114


In [70]:
table = pd.crosstab(
    cookie_model["butter_melted"],
    cookie_model["structural_crispy"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.8149749847028103)

### Preparation Methods and Redefined Crispiness

Additional baking and preparation variables were tested to determine whether they influenced structural crispiness. Bake time, oven temperature, melted butter, and creamed butter showed no statistically significant relationships with structural crispiness (all p > 0.05).

Combined with the ingredient ratio analysis, these results suggest that crispiness is not primarily determined by recipe composition or preparation technique alone. Instead, physical cookie structure appears to be the dominant factor. Characteristics such as spreading, flatness, and thinness were strongly associated with crispiness, likely because they increase surface area and promote moisture loss during baking.

In [71]:
soft_ratio_test = cookie_model.groupby("soft")[
    [
        "fat_flour_ratio",
        "sugar_flour_ratio",
        "brown_sugar_fraction"
    ]
].mean()

soft_ratio_test

,fat_flour_ratio,sugar_flour_ratio,brown_sugar_fraction
soft,,,
0,0.462211,0.815681,0.572250
1,0.470626,0.783693,0.578975


In [72]:
from scipy.stats import ttest_ind

ttest_ind(
    cookie_model.loc[cookie_model["soft"] == 1, "fat_flour_ratio"].dropna(),
    cookie_model.loc[cookie_model["soft"] == 0, "fat_flour_ratio"].dropna()
)

TtestResult(statistic=np.float64(0.22865211042824232), pvalue=np.float64(0.8192381410598665), df=np.float64(473.0))

In [73]:
ttest_ind(
    cookie_model.loc[cookie_model["soft"] == 1, "sugar_flour_ratio"].dropna(),
    cookie_model.loc[cookie_model["soft"] == 0, "sugar_flour_ratio"].dropna()
)

TtestResult(statistic=np.float64(-0.689492976157497), pvalue=np.float64(0.49085119905913754), df=np.float64(473.0))

In [74]:
ttest_ind(
    cookie_model.loc[cookie_model["soft"] == 1, "brown_sugar_fraction"].dropna(),
    cookie_model.loc[cookie_model["soft"] == 0, "brown_sugar_fraction"].dropna()
)

TtestResult(statistic=np.float64(0.29729101392145024), pvalue=np.float64(0.7663748585209526), df=np.float64(473.0))

In [75]:
pd.crosstab(
    cookie_model["brown_sugar"],
    cookie_model["soft"],
    normalize="index"
)

soft,0,1
brown_sugar,,
0.000000,0.596774,0.403226
0.250000,0.636364,0.363636
0.333333,0.666667,0.333333
0.375000,1.000000,0.000000
0.500000,0.744186,0.255814
0.666667,0.636364,0.363636
0.750000,0.670588,0.329412
1.000000,0.519608,0.480392
2.000000,0.750000,0.250000


In [76]:
from scipy.stats import chi2_contingency

table = pd.crosstab(
    cookie_model["brown_sugar"],
    cookie_model["soft"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.23389604481795054)

In [77]:
from scipy.stats import chi2_contingency

table = pd.crosstab(
    cookie_model["brown_sugar"],
    cookie_model["soft"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.23389604481795054)

In [78]:
pd.crosstab(
    cookie_model["brown_sugar"],
    cookie_model["soft"],
    normalize="index"
)

soft,0,1
brown_sugar,,
0.000000,0.596774,0.403226
0.250000,0.636364,0.363636
0.333333,0.666667,0.333333
0.375000,1.000000,0.000000
0.500000,0.744186,0.255814
0.666667,0.636364,0.363636
0.750000,0.670588,0.329412
1.000000,0.519608,0.480392
2.000000,0.750000,0.250000


In [79]:
cookie_model["has_brown_sugar"] = (
    cookie_model["brown_sugar"] > 0
).astype(int)

In [80]:
pd.crosstab(
    cookie_model["has_brown_sugar"],
    cookie_model["soft"],
    normalize="index"
)

soft,0,1
has_brown_sugar,,
0,0.596774,0.403226
1,0.636678,0.363322


In [81]:
table = pd.crosstab(
    cookie_model["has_brown_sugar"],
    cookie_model["soft"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.4364854733041308)

In [82]:
cookie_model["has_cornstarch"] = (
    cookie_model["cornstarch"] > 0
).astype(int)

In [83]:
pd.crosstab(
    cookie_model["has_cornstarch"],
    cookie_model["soft"],
    normalize="index"
)

soft,0,1
has_cornstarch,,
0,0.623126,0.376874
1,0.500000,0.500000


In [84]:
from scipy.stats import chi2_contingency

table = pd.crosstab(
    cookie_model["has_cornstarch"],
    cookie_model["soft"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.7306271803427503)

In [85]:
cookie_model["has_egg"] = (
    cookie_model["egg"] > 0
).astype(int)

In [86]:
pd.crosstab(
    cookie_model["has_egg"],
    cookie_model["soft"],
    normalize="index"
)

soft,0,1
has_egg,,
0,0.613181,0.386819
1,0.642857,0.357143


In [87]:
table = pd.crosstab(
    cookie_model["has_egg"],
    cookie_model["soft"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.630182330905469)

In [88]:
cookie_model.groupby("soft")[
    ["bake_time"]
].mean()

,bake_time
soft,
0,10.153924
1,9.900218


In [89]:
from scipy.stats import ttest_ind

ttest_ind(
    cookie_model.loc[cookie_model["soft"] == 1, "bake_time"].dropna(),
    cookie_model.loc[cookie_model["soft"] == 0, "bake_time"].dropna()
)

TtestResult(statistic=np.float64(-0.6468225254299456), pvalue=np.float64(0.5181150336963541), df=np.float64(403.0))

In [90]:
cookie_model.groupby("soft")[
    ["oven_temp"]
].mean()

,oven_temp
soft,
0,358.218954
1,356.363636


In [91]:
ttest_ind(
    cookie_model.loc[cookie_model["soft"] == 1, "oven_temp"].dropna(),
    cookie_model.loc[cookie_model["soft"] == 0, "oven_temp"].dropna()
)

TtestResult(statistic=np.float64(-0.8455645795994857), pvalue=np.float64(0.39829251375866237), df=np.float64(407.0))

In [92]:
table = pd.crosstab(
    cookie_model["butter_melted"],
    cookie_model["soft"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.23076504140013263)

In [93]:
table = pd.crosstab(
    cookie_model["butter_creamed"],
    cookie_model["soft"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.809305100515644)

In [94]:
table = pd.crosstab(
    cookie_model["chilled"],
    cookie_model["soft"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.7298916806866879)

In [95]:
table = pd.crosstab(
    cookie_model["rested"],
    cookie_model["soft"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.9705038695332504)

## Softness Analysis

Softness was analyzed using ingredient composition, preparation methods, and baking conditions to determine which factors contributed to a softer cookie texture.

Ingredient ratios, including fat/flour ratio, sugar/flour ratio, and brown sugar fraction, showed no statistically significant relationships with softness (all p > 0.05). Individual ingredients commonly associated with softer cookies, including brown sugar, cornstarch, and eggs, were also not significant predictors.

Additionally, baking and preparation variables such as bake time, oven temperature, butter method (melted or creamed), chilling, and resting did not show significant relationships with softness (all p > 0.05).

These results suggest that softness is not driven by a single ingredient or process variable within this dataset. Instead, softness may result from a combination of interacting factors, such as overall recipe balance, moisture retention, and interactions between ingredients that are not captured by individual feature testing.

In [96]:
cookie_model.groupby("thick")[
    [
        "fat_flour_ratio",
        "sugar_flour_ratio",
        "brown_sugar_fraction"
    ]
].mean()

,fat_flour_ratio,sugar_flour_ratio,brown_sugar_fraction
thick,,,
0,0.467666,0.816202,0.573347
1,0.447355,0.702893,0.586352


In [97]:
ttest_ind(
    cookie_model.loc[cookie_model["thick"] == 1, "sugar_flour_ratio"].dropna(),
    cookie_model.loc[cookie_model["thick"] == 0, "sugar_flour_ratio"].dropna()
)

TtestResult(statistic=np.float64(-1.5885329851592147), pvalue=np.float64(0.11283404847805542), df=np.float64(473.0))

In [98]:
ttest_ind(
    cookie_model.loc[cookie_model["thick"] == 1, "brown_sugar_fraction"].dropna(),
    cookie_model.loc[cookie_model["thick"] == 0, "brown_sugar_fraction"].dropna()
)

TtestResult(statistic=np.float64(0.3730753747927496), pvalue=np.float64(0.7092594769808647), df=np.float64(473.0))

In [99]:
from scipy.stats import ttest_ind

ttest_ind(
    cookie_model.loc[cookie_model["thick"] == 1, "fat_flour_ratio"].dropna(),
    cookie_model.loc[cookie_model["thick"] == 0, "fat_flour_ratio"].dropna()
)

TtestResult(statistic=np.float64(-0.35822180285677596), pvalue=np.float64(0.7203371450538645), df=np.float64(473.0))

In [100]:
table = pd.crosstab(
    cookie_model["has_cornstarch"],
    cookie_model["thick"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(1.0)

In [101]:
table = pd.crosstab(
    cookie_model["has_egg"],
    cookie_model["thick"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.2400683581000556)

In [102]:
table = pd.crosstab(
    cookie_model["shortening"],
    cookie_model["thick"]
)

chi2, p, dof, expected = chi2_contingency(table)

p

np.float64(0.40554473081537695)

In [103]:
cookie_model.groupby("thick")[
    ["bake_time"]
].mean()

,bake_time
thick,
0,10.069083
1,9.967803


In [104]:
ttest_ind(
    cookie_model.loc[cookie_model["thick"] == 1, "bake_time"].dropna(),
    cookie_model.loc[cookie_model["thick"] == 0, "bake_time"].dropna()
)

TtestResult(statistic=np.float64(-0.1656532621457933), pvalue=np.float64(0.8685128974789811), df=np.float64(403.0))

In [105]:
cookie_model.groupby("thick")[
    ["oven_temp"]
].mean()

,oven_temp
thick,
0,358.051740
1,353.222222


In [106]:
ttest_ind(
    cookie_model.loc[cookie_model["thick"] == 1, "oven_temp"].dropna(),
    cookie_model.loc[cookie_model["thick"] == 0, "oven_temp"].dropna()
)

TtestResult(statistic=np.float64(-1.4238246689722378), pvalue=np.float64(0.15526340783105527), df=np.float64(407.0))

In [107]:
for col in [
    "butter_melted",
    "butter_creamed",
    "chilled",
    "rested"
]:
    table = pd.crosstab(
        cookie_model[col],
        cookie_model["thick"]
    )

    chi2, p, dof, expected = chi2_contingency(table)

    print(col, p)

butter_melted 1.0
butter_creamed 0.4509212173852202
chilled 0.5308903538574454
rested 0.40941389010746776


## Thickness Analysis

Thickness was analyzed to determine whether ingredient composition or recipe characteristics contributed to producing thicker cookies.

Ingredient ratios showed limited evidence of influencing thickness. Fat/flour ratio and brown sugar fraction were not significantly associated with thickness (p > 0.05). Sugar/flour ratio showed the strongest trend among the tested ratios, with thick cookies having a slightly different sugar-to-flour balance, but this relationship was still not statistically significant (p = 0.113). This suggests that overall ingredient proportions alone do not strongly predict cookie thickness.

Since thickness is closely related to cookie shape and spread behavior, structural characteristics were also considered. Unlike crispiness, which was strongly associated with thinness, flatness, and spreading, thickness likely reflects reduced spreading and greater dough structure. This suggests that physical characteristics of the cookie may play a larger role in determining thickness than individual ingredient ratios.

Overall, thickness did not appear to be driven by a single ingredient ratio. Instead, thickness is likely influenced by a combination of structural factors that affect how much a cookie spreads during baking.

## Texture Analysis Summary

### Chewy Cookies
Chewiness appeared to be most influenced by ingredients related to moisture retention. Brown sugar showed the strongest relationship, with cookies containing brown sugar having a higher proportion of chewy cookies. This suggests that ingredients affecting moisture and texture may play a larger role in creating a chewy cookie.

### Crispy Cookies
Crispiness was primarily associated with cookie structure rather than ingredient composition. After redefining crispiness as a structural category (crispy, thin, flat, or spread), ingredient ratios, bake time, oven temperature, and butter methods were not significant predictors. Instead, thinness, flatness, and spreading were the strongest characteristics associated with crispy cookies, likely due to increased surface area and moisture loss.

### Soft Cookies
Softness did not have a single dominant predictor in this dataset. Ingredient ratios, individual ingredients such as brown sugar and cornstarch, and baking variables including bake time, oven temperature, chilling, and butter method were not statistically significant. This suggests that softness likely results from the interaction of multiple recipe factors rather than one specific variable.

### Thick Cookies
Thickness was also not strongly explained by ingredient ratios. Fat/flour ratio and brown sugar fraction showed no significant relationship, while sugar/flour ratio showed the closest trend but remained statistically insignificant (p = 0.113). This suggests that thickness is likely driven more by structural factors affecting spread and shape rather than ingredient proportions alone.

### Overall Findings
Different cookie textures appear to be driven by different mechanisms:
- Chewy cookies → moisture retention and ingredient composition
- Crispy cookies → physical structure and spreading behavior
- Soft cookies → combination of interacting factors
- Thick cookies → reduced spreading and structural characteristics

In [108]:
cookie_model.groupby("crispy")[
[
"thin",
"flat",
"spread",
"thick"
]
].mean()

,thin,flat,spread,thick
crispy,,,,
0,0.391192,0.181347,0.113990,0.075130
1,0.707865,0.449438,0.370787,0.269663


In [109]:
cookie_model.groupby("thick")[
[
"thin",
"flat",
"spread",
"crispy"
]
].mean()

,thin,flat,spread,crispy
thick,,,,
0,0.409953,0.199052,0.111374,0.154028
1,0.773585,0.490566,0.566038,0.452830


In [110]:
cookie_model.groupby("structural_crispy")[
[
"thin",
"flat",
"spread",
"thick"
]
].mean()

,thin,flat,spread,thick
structural_crispy,,,,
0,0.000000,0.000000,0.000000,0.024510
1,0.789668,0.405904,0.284133,0.177122


## Crispy vs. Thick: Structural Differences

Crispiness appeared to be primarily associated with cookie structure. Crispy cookies were substantially more likely to be labeled thin (70.8% vs. 39.1%), flat (44.9% vs. 18.1%), and spread (37.1% vs. 11.4%). These characteristics increase surface area and promote moisture loss, providing a likely explanation for the relationship between structure and crisp texture.

When redefining crispiness as a structural category (crispy, thin, flat, or spread), the relationship became even stronger. Structural crispy cookies were labeled thin 79.0% of the time, compared with none of the non-structural crispy cookies. This suggests that thinness is the strongest structural indicator of crispiness.

However, texture labels were not mutually exclusive, as some cookies were simultaneously classified as thick and crispy. This suggests that thickness alone does not determine final texture; rather, the interaction between cookie structure and moisture retention likely influences the final outcome.

## Logistic Regression

In [111]:
from sklearn.metrics import accuracy_score, classification_report

log_accuracy = accuracy_score(
    y_test,
    log_predictions
)

log_accuracy

NameError: name 'y_test' is not defined

In [ ]:
print(classification_report(y_test, log_predictions))

              precision    recall  f1-score   support

           0       0.74      0.96      0.84       277
           1       0.79      0.33      0.47       138

    accuracy                           0.75       415
   macro avg       0.77      0.65      0.65       415
weighted avg       0.76      0.75      0.71       415



## Chewy Texture Prediction: Logistic Regression

A logistic regression model was used to predict whether a cookie would be classified as chewy based on ingredient composition, preparation methods, and recipe ratios. The model achieved an accuracy of 74.9%.

The model performed well at identifying non-chewy cookies, achieving a recall of 96% for class 0. However, it struggled to identify chewy cookies, with a recall of 33%. While predictions labeled as chewy were generally accurate (79% precision), many actual chewy cookies were missed.

This suggests that chewiness is influenced by multiple interacting factors rather than a single dominant ingredient or process variable. The results align with the statistical analysis, where brown sugar showed an association with chewiness, but other variables contributed to the final texture.

## Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(
    random_state=42
)

rf.fit(
    X_train,
    y_train
)

rf_predictions = rf.predict(X_test)

rf_accuracy = accuracy_score(
    y_test,
    rf_predictions
)

rf_accuracy

0.8698795180722891

In [ ]:
print(classification_report(y_test, rf_predictions))

              precision    recall  f1-score   support

           0       0.88      0.94      0.91       277
           1       0.85      0.74      0.79       138

    accuracy                           0.87       415
   macro avg       0.86      0.84      0.85       415
weighted avg       0.87      0.87      0.87       415



In [ ]:
importance = pd.DataFrame({
    "feature": X.columns,
    "importance": rf.feature_importances_
})

importance.sort_values(
    "importance",
    ascending=False
)

,feature,importance
13,bake_time,0.372536
14,oven_temp,0.164063
7,baking_powder,0.083915
12,butter_creamed,0.076172
1,shortening,0.053315
3,granulated_sugar,0.052730
0,butter,0.045004
9,chilled,0.033782
6,baking_soda,0.026342
2,brown_sugar,0.025727


## Chewy Texture Prediction

Two classification models were used to predict chewy cookies: logistic regression and random forest. Logistic regression achieved 74.9% accuracy but struggled to identify chewy cookies, with only 33% recall. A random forest model improved performance to 87.0% accuracy and increased chewy-cookie recall to 74%, suggesting that chewiness is influenced by nonlinear interactions between multiple recipe features.

The random forest identified bake time (37.3%) and oven temperature (16.4%) as the strongest predictors, followed by baking powder, creamed butter, and shortening. While brown sugar showed a relationship with chewiness in statistical testing, the machine learning model suggests that chewiness is better explained by a combination of ingredients and baking conditions rather than a single ingredient.

In [ ]:
y = cookie_model["structural_crispy"]

In [ ]:
X = cookie_model[features]

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [ ]:
X.isna().sum().sort_values(ascending=False)

bake_time           70
chilled             66
rested              66
butter_melted       66
butter_creamed      66
oven_temp           66
butter               0
shortening           0
brown_sugar          0
granulated_sugar     0
egg                  0
cornstarch           0
baking_soda          0
baking_powder        0
vanilla              0
dtype: int64

In [ ]:
X = X.fillna(X.median())

In [ ]:
X.isna().sum().sum()

np.int64(0)

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

log_model = LogisticRegression(
    max_iter=1000
)

log_model.fit(
    X_train,
    y_train
)

log_predictions = log_model.predict(X_test)

log_accuracy = accuracy_score(
    y_test,
    log_predictions
)

print(log_accuracy)

print(
    classification_report(
        y_test,
        log_predictions
    )
)

0.4842105263157895
              precision    recall  f1-score   support

           0       0.00      0.00      0.00        43
           1       0.52      0.88      0.65        52

    accuracy                           0.48        95
   macro avg       0.26      0.44      0.33        95
weighted avg       0.28      0.48      0.36        95



In [ ]:
log_coefficients = pd.DataFrame({
    "feature": X.columns,
    "coefficient": log_model.coef_[0]
})

log_coefficients.sort_values(
    "coefficient",
    ascending=False
)

,feature,coefficient
5,cornstarch,0.568155
10,rested,0.173349
7,baking_powder,0.105528
0,butter,0.072608
12,butter_creamed,0.069772
4,egg,0.063003
9,chilled,0.023133
14,oven_temp,0.000329
13,bake_time,-0.022338
8,vanilla,-0.029507


Before modeling, missing numerical features were handled using median imputation. This preserved the full dataset while reducing the influence of outliers compared with mean imputation.

## Structural Crispiness Prediction: Logistic Regression

A logistic regression model was used to determine whether ingredient composition and baking variables could predict structural crispiness. The model achieved 48.4% accuracy and struggled to distinguish between structural crispy and non-structural crispy cookies.

The model primarily predicted cookies as structural crispy, resulting in high recall (88%) but poor performance identifying non-structural crispy cookies. This suggests that ingredient composition and baking variables alone are insufficient predictors of crispiness.

These results align with the statistical analysis, where crispiness was most strongly associated with structural characteristics such as thinness, flatness, and spreading rather than individual ingredients or baking conditions.

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(
    random_state=42
)

rf_model.fit(
    X_train,
    y_train
)

rf_predictions = rf_model.predict(X_test)

print(accuracy_score(y_test, rf_predictions))

print(
    classification_report(
        y_test,
        rf_predictions
    )
)

0.6105263157894737
              precision    recall  f1-score   support

           0       0.57      0.56      0.56        43
           1       0.64      0.65      0.65        52

    accuracy                           0.61        95
   macro avg       0.61      0.61      0.61        95
weighted avg       0.61      0.61      0.61        95



In [ ]:
rf_importance = pd.DataFrame({
    "feature": X.columns,
    "importance": rf_model.feature_importances_
})

rf_importance.sort_values(
    "importance",
    ascending=False
)

,feature,importance
13,bake_time,0.291017
2,brown_sugar,0.144135
0,butter,0.104314
14,oven_temp,0.098369
3,granulated_sugar,0.097513
4,egg,0.047782
1,shortening,0.043368
12,butter_creamed,0.041502
7,baking_powder,0.039865
9,chilled,0.025629


## Structural Crispiness Prediction

A logistic regression and random forest model were used to predict structural crispiness from ingredient composition and baking variables.

Logistic regression performed poorly, achieving 48.4% accuracy and primarily predicting cookies as structural crispy. A random forest model improved performance to 61.1% accuracy, with more balanced classification across both groups.

The random forest identified bake time, brown sugar, butter, oven temperature, and granulated sugar as the most influential features. However, the moderate prediction accuracy suggests that ingredient composition and baking conditions alone are insufficient for reliably predicting crispiness.

These findings reinforce the statistical analysis: crispiness appears to be primarily driven by physical cookie structure, particularly thinness, flatness, and spreading, which were not included as predictive features. Recipe chemistry influences texture indirectly through its effect on cookie structure.

In [ ]:
y = cookie_model["soft"]

X = cookie_model[features]

In [ ]:
X = X.fillna(X.median())

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

log_model = LogisticRegression(
    max_iter=1000
)

log_model.fit(
    X_train,
    y_train
)

log_predictions = log_model.predict(X_test)

log_accuracy = accuracy_score(
    y_test,
    log_predictions
)

print(log_accuracy)

print(
    classification_report(
        y_test,
        log_predictions
    )
)

0.6526315789473685
              precision    recall  f1-score   support

           0       0.66      0.98      0.79        63
           1       0.00      0.00      0.00        32

    accuracy                           0.65        95
   macro avg       0.33      0.49      0.39        95
weighted avg       0.44      0.65      0.52        95



In [ ]:
log_coefficients = pd.DataFrame({
    "feature": X.columns,
    "coefficient": log_model.coef_[0]
})

log_coefficients.sort_values(
    "coefficient",
    ascending=False
)

,feature,coefficient
11,butter_melted,0.691465
6,baking_soda,0.432072
5,cornstarch,0.282687
1,shortening,0.159394
10,rested,0.138829
7,baking_powder,0.122916
8,vanilla,0.066009
4,egg,0.046465
0,butter,0.029908
14,oven_temp,-0.004827


## Soft Texture Prediction: Logistic Regression

A logistic regression model was used to predict soft cookies using ingredient composition and baking variables. The model achieved 65.3% accuracy; however, performance was driven primarily by correctly identifying non-soft cookies. The model failed to identify soft cookies, achieving 0% recall for the soft class.

This suggests that softness cannot be reliably predicted using the selected recipe features alone. Although melted butter, baking soda, and cornstarch showed positive coefficients, the weak classification performance indicates that softness likely emerges from complex interactions between ingredients, moisture retention, and baking conditions.

These results align with the statistical analysis, where no individual ingredient, ratio, or baking variable showed a significant relationship with softness.

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(
    random_state=42
)

rf_model.fit(
    X_train,
    y_train
)

rf_predictions = rf_model.predict(X_test)

rf_accuracy = accuracy_score(
    y_test,
    rf_predictions
)

print(rf_accuracy)

print(
    classification_report(
        y_test,
        rf_predictions
    )
)

0.6105263157894737
              precision    recall  f1-score   support

           0       0.68      0.79      0.73        63
           1       0.38      0.25      0.30        32

    accuracy                           0.61        95
   macro avg       0.53      0.52      0.52        95
weighted avg       0.58      0.61      0.59        95



In [ ]:
rf_importance = pd.DataFrame({
    "feature": X.columns,
    "importance": rf_model.feature_importances_
})

rf_importance.sort_values(
    "importance",
    ascending=False
)

,feature,importance
13,bake_time,0.288905
2,brown_sugar,0.134511
0,butter,0.112510
14,oven_temp,0.095559
3,granulated_sugar,0.092000
4,egg,0.056463
12,butter_creamed,0.048865
1,shortening,0.045259
7,baking_powder,0.032833
9,chilled,0.028241


## Soft Texture Prediction

Logistic regression and random forest models were used to predict soft cookies using ingredient composition and baking variables.

The logistic regression model achieved 65.3% accuracy but failed to identify soft cookies, suggesting that linear relationships between individual features were insufficient for prediction. The random forest model improved classification balance, achieving 61.1% accuracy and identifying some soft cookies (25% recall), but overall predictive performance remained limited.

The random forest identified bake time, brown sugar, butter, oven temperature, and granulated sugar as the most influential features. However, the moderate accuracy and low recall for soft cookies suggest that softness is likely determined by complex interactions between ingredients and baking conditions rather than a single dominant factor.

These results align with the statistical analysis, where no individual ingredient, ratio, or preparation variable showed a significant relationship with softness.

In [ ]:
y = cookie_model["thick"]

X = cookie_model[features]

X = X.fillna(X.median())

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

log_model = LogisticRegression(
    max_iter=1000
)

log_model.fit(
    X_train,
    y_train
)

log_predictions = log_model.predict(X_test)

log_accuracy = accuracy_score(
    y_test,
    log_predictions
)

print(log_accuracy)

print(
    classification_report(
        y_test,
        log_predictions
    )
)

0.9263157894736842
              precision    recall  f1-score   support

           0       0.93      1.00      0.96        88
           1       0.00      0.00      0.00         7

    accuracy                           0.93        95
   macro avg       0.46      0.50      0.48        95
weighted avg       0.86      0.93      0.89        95



/Users/abbygoblick/cookie-lab/.venv/lib/python3.14/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/abbygoblick/cookie-lab/.venv/lib/python3.14/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/abbygoblick/cookie-lab/.venv/lib/python3.14/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metri

In [ ]:
log_coefficients = pd.DataFrame({
    "feature": X.columns,
    "coefficient": log_model.coef_[0]
})

log_coefficients.sort_values(
    "coefficient",
    ascending=False
)

,feature,coefficient
7,baking_powder,0.335097
12,butter_creamed,0.284901
6,baking_soda,0.254688
5,cornstarch,0.136975
8,vanilla,0.087525
14,oven_temp,-0.011504
13,bake_time,-0.018089
11,butter_melted,-0.029510
10,rested,-0.344395
2,brown_sugar,-0.381390


## Thick Texture Prediction: Logistic Regression

A logistic regression model was used to predict thick cookies based on ingredient composition and baking variables. The model achieved 92.6% accuracy; however, this performance was driven by the strong imbalance between classes, with only seven thick cookies appearing in the test set.

The model successfully classified non-thick cookies but failed to identify any thick cookies, resulting in 0% recall for the thick class. This indicates that recipe composition alone was insufficient for reliably predicting thickness.

Although baking powder, creamed butter, baking soda, and cornstarch showed positive coefficients, these relationships should be interpreted cautiously due to the limited number of thick examples. Additional structural features, such as cookie height, diameter, and spread behavior, would likely improve thickness prediction.

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(
    random_state=42
)

rf_model.fit(
    X_train,
    y_train
)

rf_predictions = rf_model.predict(X_test)

rf_accuracy = accuracy_score(
    y_test,
    rf_predictions
)

print(rf_accuracy)

print(
    classification_report(
        y_test,
        rf_predictions
    )
)

0.8736842105263158
              precision    recall  f1-score   support

           0       0.92      0.94      0.93        88
           1       0.00      0.00      0.00         7

    accuracy                           0.87        95
   macro avg       0.46      0.47      0.47        95
weighted avg       0.85      0.87      0.86        95



In [ ]:
rf_importance = pd.DataFrame({
    "feature": X.columns,
    "importance": rf_model.feature_importances_
})

rf_importance.sort_values(
    "importance",
    ascending=False
)

,feature,importance
13,bake_time,0.312216
2,brown_sugar,0.146547
0,butter,0.115569
14,oven_temp,0.087669
4,egg,0.065333
3,granulated_sugar,0.062408
12,butter_creamed,0.048409
7,baking_powder,0.041850
1,shortening,0.036881
9,chilled,0.021698


## Thick Texture Prediction

Both logistic regression and random forest models struggled to predict thick cookies. Although accuracy was high (92.6% logistic regression; 87.4% random forest), this was primarily due to class imbalance, as only seven thick cookies appeared in the test set. Neither model successfully identified thick cookies, resulting in 0% recall for the thick class.

The random forest identified bake time, brown sugar, butter, and oven temperature as the most influential features, but the inability to classify thick cookies suggests that recipe composition alone does not capture the factors driving thickness.

These results support the earlier statistical analysis, where ingredient ratios showed limited relationships with thickness. Similar to crispiness, thickness appears to be primarily influenced by structural characteristics of the final cookie rather than ingredient composition alone.

# Final Analysis Summary

## Statistical Findings

- **Chewy**
  - Brown sugar showed the strongest relationship with chewiness.
  - Suggests moisture retention is important, but chewiness likely results from multiple interacting factors.

- **Crispy**
  - Strongly associated with **thinness, flatness, and spreading**.
  - Ingredient ratios, bake time, oven temperature, and butter methods were not significant.
  - Suggests crispiness is driven more by **cookie structure** than recipe composition.

- **Soft**
  - No individual ingredient, ratio, or baking variable was significant.
  - Suggests softness is a **multifactorial texture outcome**.

- **Thick**
  - Ingredient ratios were not significant predictors.
  - Sugar/flour ratio showed the closest trend (**p = 0.113**) but was not significant.
  - Suggests thickness depends more on structural characteristics.

---

# Machine Learning Results

Models:
- **Logistic Regression** → interpretable feature relationships
- **Random Forest** → captures nonlinear interactions

| Texture | Logistic Regression | Random Forest |
|---|---:|---:|
| Chewy | 74.9% | **87.0%** |
| Structural Crispy | 48.4% | **61.1%** |
| Soft | **65.3%** | 61.1% |
| Thick | 92.6%* | 87.4%* |

\*High accuracy due to class imbalance; models failed to identify thick cookies.

---

## Model Insights

### Chewy
- Best predicted texture.
- Random Forest improved accuracy from **74.9% → 87.0%**.
- Top predictors:
  - **Bake time (37.3%)**
  - **Oven temperature (16.4%)**
  - **Baking powder (8.4%)**
  - **Creamed butter (7.6%)**
- Suggests chewiness depends on interactions between ingredients and baking conditions.

### Structural Crispy
- Recipe-based prediction was limited.
- Random Forest improved accuracy from **48.4% → 61.1%**.
- Top predictors:
  - **Bake time (29.1%)**
  - **Brown sugar (14.4%)**
  - **Butter (10.4%)**
  - **Oven temperature (9.8%)**
- Supports finding that crispiness is mainly driven by **physical structure** (thin, flat, spread).

### Soft
- Difficult to predict using recipe features.
- Logistic Regression: **65.3%**
- Random Forest: **61.1%**
- Top Random Forest predictors:
  - **Bake time (28.9%)**
  - **Brown sugar (13.5%)**
  - **Butter (11.3%)**
- Suggests softness results from complex interactions rather than one dominant factor.

### Thick
- Accuracy was misleading due to limited examples.
- Logistic Regression: **92.6%**
- Random Forest: **87.4%**
- Both models had **0% recall** for thick cookies.
- Top Random Forest predictors:
  - **Bake time (31.2%)**
  - **Brown sugar (14.7%)**
  - **Butter (11.6%)**
- Suggests thickness requires additional structural measurements beyond recipe data.

---

# Overall Conclusion

- Cookie texture is not determined by a single ingredient.
- Different textures are driven by different mechanisms:
  - **Chewy →** moisture retention and ingredient interactions
  - **Crispy →** structural properties like thinness and spreading
  - **Soft →** complex interactions between many factors
  - **Thick →** dough structure and spread behavior

Recipe composition influences texture, but the final cookie outcome is strongly shaped by how the dough behaves during baking.

In [113]:
from sklearn.ensemble import RandomForestClassifier
import joblib
import os

os.makedirs("models", exist_ok=True)


features = [
    "butter",
    "shortening",
    "brown_sugar",
    "granulated_sugar",
    "egg",
    "cornstarch",
    "baking_soda",
    "baking_powder",
    "vanilla",
    "chilled",
    "rested",
    "butter_melted",
    "butter_creamed",
    "bake_time",
    "oven_temp"
]


X = cookie_model[features]
X = X.fillna(X.median())

y = cookie_model["chewy"]


from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)


chewy_rf = RandomForestClassifier(
    random_state=42
)

chewy_rf.fit(
    X_train,
    y_train
)


joblib.dump(
    chewy_rf,
    "models/chewy_random_forest.pkl"
)

['models/chewy_random_forest.pkl']

In [114]:
y = cookie_model["structural_crispy"]


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)


crispy_rf = RandomForestClassifier(
    random_state=42
)

crispy_rf.fit(
    X_train,
    y_train
)


joblib.dump(
    crispy_rf,
    "models/structural_crispy_random_forest.pkl"
)

['models/structural_crispy_random_forest.pkl']

In [115]:
y = cookie_model["soft"]


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)


soft_rf = RandomForestClassifier(
    random_state=42
)

soft_rf.fit(
    X_train,
    y_train
)


joblib.dump(
    soft_rf,
    "models/soft_random_forest.pkl"
)

['models/soft_random_forest.pkl']

In [116]:
y = cookie_model["thick"]


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)


thick_rf = RandomForestClassifier(
    random_state=42
)

thick_rf.fit(
    X_train,
    y_train
)


joblib.dump(
    thick_rf,
    "models/thick_random_forest.pkl"
)

['models/thick_random_forest.pkl']

In [117]:
import os

os.listdir("models")

['soft_random_forest.pkl',
 'chewy_random_forest.pkl',
 'structural_crispy_random_forest.pkl',
 'thick_random_forest.pkl']